### PR TITLE
Speed improvements for StackedChainNER

### PR DESCRIPTION
**Pull request by Craigacp**
This set of commits reduces the runtime of the StackedChainNER on an example document with about 8300 words from about 1000ms to about 667ms (using a lightly processed version of Barack Obama's wikipedia page as the target document, averaged across 50 runs). This requires no change on the user's part, and the current serialized models are compatible.

* It adds a new method to TriePhraseLexicon to let it reuse the lemmatised form of a token (as previously it lemmatised it once for each lexicon), and shifts StackedChainNer to use that method.
* It merges the three for loops over tokens at the end of StackedChainNer.initFeatures into a single for loop, and the three for loops at the start of StackedChainNer.initSecondaryFeatures into a single for loop.
* It adds an extra CircularBuffer class to cc.factorie.util (because I couldn't figure out an appropriate Scala collection with the right performance characteristics) which is the top n most recent elements held in an array. It's useful for replacing the tokens.foreach(t => t.prevwindow... ) operation that is common in the NER, to save doing a potentially complicated operation many times.
* initFeatures is modified to use the CircularBuffer to store the feature added to NEXTWINDOW & PREVWINDOW so simplifyDigits(token).toLowercase is called once per token, instead of 8 times as in the previous code. It's not used on the initSecondaryFeatures prev/next window features as that operation isn't expensive enough to justify the extra code IMO, and it does make it harder to figure out what's happening.
* The call to addNeighboringFeatureConjunctions was swapped to a new method addNeighboringFeaturesFast. The addNeighboringFeatureConjunctions method has lots of extra logic which is unnecessary when called with the arguments used by StackedChainNer (as it doesn't make conjunctions, only adds the neighbouring features which match the regex). addNeighboringFeatureConjunctions is supposed to subsume the earlier addNeighboringFeatures method, but the behaviour of the two of them has diverged (I added a comment noting some of this divergence), so the new addNeighboringFeaturesFast follows addNeighboringFeatureConjunctions to preserve compatibility with existing trained models. addNeighboringFeatures should probably be deprecated, it has no usages inside of factorie.

### CODE REVIEW

File: src/main/scala/cc/factorie/util/CircularBuffer.scala [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+  def flush() : Unit = {
+    var i = 0
+    while (i < size) {
+      buffer(i) = null.asInstanceOf[A]
+      i += 1
+    }
+  }
+
+  def add(item : A) : A = {
```
Comment (strubell) [inline src/main/scala/cc/factorie/util/CircularBuffer.scala]: So this is super nitpicky, but would it be too much trouble to implement this data structure more Scala-istically using += for add and apply for get? :)

**Comment by strubell** [comment]
This looks good to me. I support deprecation of addNeighboringFeatures.

File: src/main/scala/cc/factorie/util/CircularBuffer.scala (hunk 1 above)
Comment (Craigacp) [inline src/main/scala/cc/factorie/util/CircularBuffer.scala]: Sure.

**Comment by andrewmccallum** [comment]
Can we then just remove it, and rename addNeighboringFeaturesFast to take its place?

**Comment by Craigacp** [comment]
Sure. I try and avoid changing the semantics of methods when I make these patches to ensure I don't break anyone else's code. I can swap the old one out if you want.

**Comment by Craigacp** [comment]
I have a commit which renames the method and removes the old one if you want me to push that?

**Comment by strubell** [comment]
@Craigacp go ahead and rename/remove

**Comment by Craigacp** [comment]
Done.

**Comment by strubell** [comment]
:+1:

**Pull request merged**